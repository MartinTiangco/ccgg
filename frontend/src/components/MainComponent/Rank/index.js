import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import { Grid, Typography } from "@material-ui/core";

import emblems from "../../../images";

// if the player does not have a solo or a flex rank
const defaultStats = {
  queueType: "",
  tier: "UNRANKED",
  rank: "",
  leaguePoints: 0,
  wins: 0,
  losses: 0,
};

const useStyles = makeStyles(() => ({
  img: {
    filter: "drop-shadow(0px 5px 5px)",
  },
  imgSolo: {
    width: "70%",
  },
  imgFlex: {
    width: "50%",
  },
  heading: {
    color: "#ECF0F1",
    fontWeight: "bold",
  },
  headingSolo: {
    fontSize: "1.2rem",
  },
  headingFlex: {
    fontSize: "1rem",
  },
  textFlex: {
    fontSize: "0.8rem",
  },
}));

// Component showing the Rank of a player in either Ranked Solo or Ranked Flex modes.
const Rank = ({ isFlex, rank: rankStatus }) => {
  let statsToShow = defaultStats;
  const classes = useStyles();

  if (rankStatus && rankStatus.length !== 0) {
    const [rankedSolo, rankedFlex] = rankStatus;
    if (isFlex) {
      statsToShow = rankedFlex;
    } else {
      statsToShow = rankedSolo;
    }
  }

  const {
    queueType,
    tier,
    rank: division,
    leaguePoints,
    wins,
    losses,
  } = statsToShow;

  const isRanked = tier !== defaultStats.tier;
  const winRatio = isRanked ? Math.round((wins / (wins + losses)) * 100) : 0;

  // img element
  const imgSrc = emblems[`emblem${tier}`];
  let imgClassNames = classes.img;
  // smaller emblem img for flex
  imgClassNames +=
    queueType === "RANKED_SOLO_5x5"
      ? ` ${classes.imgSolo}`
      : ` ${classes.imgFlex}`;

  // rank heading class names
  let headingClassNames = classes.heading;
  headingClassNames +=
    queueType === "RANKED_SOLO_5x5"
      ? ` ${classes.headingSolo}`
      : ` ${classes.headingFlex}`;

  // text displays
  const testClassNames = isFlex && ` ${classes.textFlex}`; // smaller font size for flex
  const displayQueueType = isFlex ? "Ranked Flex" : "Ranked Solo";
  const displayRankLP = isRanked
    ? `${tier} ${division} / ${leaguePoints} LP`
    : "Unranked";
  const displayWR = `${winRatio}% WR | ${wins + losses} games`;

  return (
    <Grid container>
      <Grid container justify="center" alignItems="center" item xs={5}>
        <img className={imgClassNames} src={imgSrc} alt={tier} />
      </Grid>
      <Grid container item xs={7} direction="column" justify="center">
        <Typography align="center" variant="body2">
          {displayQueueType}
        </Typography>
        <Typography className={headingClassNames} align="center" variant="h5">
          {displayRankLP}
        </Typography>
        {isRanked && (
          <Typography className={testClassNames} align="center" variant="body2">
            {displayWR}
          </Typography>
        )}
      </Grid>
    </Grid>
  );
};

export default Rank;
