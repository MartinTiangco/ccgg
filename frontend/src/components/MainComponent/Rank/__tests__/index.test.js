import renderer from "react-test-renderer";
import React from "react";
import Rank from "../index";

const testDataUnranked = undefined;
const testData = [
  {
    queueType: "RANKED_SOLO_5x5",
    tier: "GOLD",
    rank: "I",
    summonerId: "EXXJEL_O4q_DiFvzHjjjw7NFi957AvQkb14Nzf7rBq-Ung",
    summonerName: "AU Dogs",
    leaguePoints: 15,
    wins: 205,
    losses: 218,
    veteran: false,
    inactive: false,
    freshBlood: false,
    hotStreak: false,
  },
  {
    queueType: "RANKED_FLEX_SR",
    tier: "SILVER",
    rank: "I",
    summonerId: "EXXJEL_O4q_DiFvzHjjjw7NFi957AvQkb14Nzf7rBq-Ung",
    summonerName: "AU Dogs",
    leaguePoints: 77,
    wins: 193,
    losses: 204,
    veteran: false,
    inactive: false,
    freshBlood: true,
    hotStreak: true,
  },
];

it("with no ranked info", () => {
  const tree = renderer.create(<Rank rank={testDataUnranked} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("with ranked solo info", () => {
  const tree = renderer.create(<Rank rank={testData} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("with ranked flex info", () => {
  const tree = renderer.create(<Rank rank={testData} isFlex />).toJSON();
  expect(tree).toMatchSnapshot();
});
