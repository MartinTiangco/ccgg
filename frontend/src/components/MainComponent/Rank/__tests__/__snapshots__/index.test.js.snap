// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with no ranked info 1`] = `
<div
  className="MuiGrid-root MuiGrid-container"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-5"
  >
    <img
      alt="UNRANKED"
      className="makeStyles-img-1 makeStyles-imgFlex-3"
      src="Emblem_Unranked.png"
    />
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-justify-xs-center MuiGrid-grid-xs-7"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
    >
      Ranked Solo
    </p>
    <h5
      className="MuiTypography-root makeStyles-heading-4 makeStyles-headingFlex-6 MuiTypography-h5 MuiTypography-alignCenter"
    >
      Unranked
    </h5>
  </div>
</div>
`;

exports[`with ranked flex info 1`] = `
<div
  className="MuiGrid-root MuiGrid-container"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-5"
  >
    <img
      alt="SILVER"
      className="makeStyles-img-1 makeStyles-imgFlex-3"
      src="Emblem_Silver.png"
    />
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-justify-xs-center MuiGrid-grid-xs-7"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
    >
      Ranked Flex
    </p>
    <h5
      className="MuiTypography-root makeStyles-heading-4 makeStyles-headingFlex-6 MuiTypography-h5 MuiTypography-alignCenter"
    >
      SILVER I / 77 LP
    </h5>
    <p
      className="MuiTypography-root  makeStyles-textFlex-7 MuiTypography-body2 MuiTypography-alignCenter"
    >
      49% WR | 397 games
    </p>
  </div>
</div>
`;

exports[`with ranked solo info 1`] = `
<div
  className="MuiGrid-root MuiGrid-container"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-center MuiGrid-grid-xs-5"
  >
    <img
      alt="GOLD"
      className="makeStyles-img-1 makeStyles-imgSolo-2"
      src="Emblem_Gold.png"
    />
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-justify-xs-center MuiGrid-grid-xs-7"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
    >
      Ranked Solo
    </p>
    <h5
      className="MuiTypography-root makeStyles-heading-4 makeStyles-headingSolo-5 MuiTypography-h5 MuiTypography-alignCenter"
    >
      GOLD I / 15 LP
    </h5>
    <p
      className="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
    >
      48% WR | 423 games
    </p>
  </div>
</div>
`;
